In [1]:
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import os
import re
from IPython.display import clear_output

------
<a id="indice"></a>

# Índice

1. **[Coleta](#coleta)**
    * [Twitter Archive](#coleta:twitter-archive)
    * [Image Prediction](#coleta:image-prediction)
    * [Twitter API](#coleta:twitter-api)
    
    
2. **[Avaliação](#avaliacao)**
    * [Twitter Archive](#avaliacao:twitter-archive)
    * [Image Prediction](#avaliacao:image-prediction)
    * [Twitter API](#avaliacao:twitter-api)
    * [Notas da avaliação](#avaliacao:anotacoes)
    
    
3. **[Limpeza](#limpeza)**
    * [Definição](#limpeza:definicao)
    * [Twitter Archive](#limpeza:twitter-archive)
    * [Image Prediction](#limpeza:image-prediction)
    * [Twitter API](#limpeza:twitter-api)


4. **[Armazenamento](#armazenamento)**

------
<a id="coleta"></a>

# Coleta

<a id="coleta:twitter-archive"></a>

## Coleta: Twitter archive

In [2]:
df_twitter_arc = pd.read_csv('data/twitter-archive-enhanced.csv')
df_twitter_arc.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
516   810984652412424192                    NaN                  NaN   
1768  678389028614488064                    NaN                  NaN   

                      timestamp  \
516   2016-12-19 23:06:23 +0000   
1768  2015-12-20 01:38:42 +0000   

                                                 source  \
516   <a href="http://twitter.com/download/iphone" r...   
1768  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
1768  This is Bella. She just learned that her final...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
516                        NaN                        NaN   
1768                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
516   https://www.gofundme.com/sams-smile,https://tw...                24   
1768  https://twitter.com/dog_rates/status/678389028...                11   

      rating_denominator   name doggo floofer  pupper puppo  
516                    7    Sam  None    None    None  None  
1768                  10  Bella  None    None  pupper  None

<a id="coleta:image-prediction"></a>

## Coleta: Image prediction

In [3]:
df_prediction = None

r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

if r.status_code is 200:
    df_prediction = pd.read_csv(pd.compat.StringIO(r.text), sep='\t')    
else:    
    print('ERROR: Image prediction request returned {status_code} status code.'.format(status_code = r.status_code))

In [4]:
df_prediction.sample(2)

tweet_id                                          jpg_url  \
1569  794355576146903043  https://pbs.twimg.com/media/CvJCabcWgAIoUxW.jpg   
1016  709901256215666688  https://pbs.twimg.com/media/CdoTbL_XIAAitq2.jpg   

      img_num              p1   p1_conf  p1_dog                p2   p2_conf  \
1569        1  cocker_spaniel  0.500509    True  golden_retriever  0.272734   
1016        2             bib  0.998814   False      handkerchief  0.000512   

      p2_dog             p3   p3_conf  p3_dog  
1569    True  jigsaw_puzzle  0.041476   False  
1016   False       umbrella  0.000224   False

<a id="coleta:twitter-api"></a>

## Coleta: Twitter API

In [5]:
# Carregar configuração do Twitter App
with open('twitter_config.json', 'r', encoding='utf-8') as file:
    app_config = json.load(file)

In [6]:
# Atribuição da configurações a variaveis locais
api_key = app_config['api_key']
api_secret = app_config['api_secret']
access_token = app_config['access_token']
access_secret = app_config['access_secret']

In [7]:
# Autentificação ao Twitter App
auth = tweepy.OAuthHandler(api_key, api_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

Buscar dados do histórico de Tweets na API

In [8]:
json_path = 'data/tweet_json.txt'
error_log = 'data/tweet_error.log'

# Salvar Tweets localmente
if(not os.path.isfile(json_path)):
    print('Aguarde. Dado o volume de requisições, a coleta de tweets pode demorar alguns minutos.')
    tweet_count = 0
    for tweet_id in df_twitter_arc.tweet_id:
        print('{percent}%'.format(percent=int((++tweet_count/df_twitter_arc.tweet_id.size)*100)))
        try:
            status = api.get_status(tweet_id)
            with open(json_path, 'a', newline='\n') as file:
                file.write(f'{json.dumps(status._json)}\n')

        except Exception as err:
            with open(error_log, 'a', newline='\n') as log:
                log.write(f'{str(tweet_id)}: {str(err.args[0])}\n')
            print(f'{str(tweet_id)}: {str(err.args[0])}')
        
        clear_output(wait=True)
else:
    print('Dados já salvos em disco, não serão executadas novas requisições à API do Twitter.')
        

Dados já salvos em disco, não serão executadas novas requisições à API do Twitter.


In [9]:
# Criar uma lista de Dictionaries com os Tweets carregados
tweets = []

with open(json_path, 'r') as file:
    for line in file:
        try:
            tweet = json.loads(line)
            
            if(tweet.get('entities', False)):
                if(tweet['entities'].get('media', False) and tweet['entities']['media'][0].get('media_url', False)):
                    tweets.append({
                        'id': int(tweet['id']),
                        'created_at': tweet['created_at'],
                        'in_reply_to_status_id': str(tweet['in_reply_to_status_id']),
                        'in_reply_to_user_id': str(tweet['in_reply_to_user_id']),
                        'is_quote_status': tweet['is_quote_status'],
                        'retweet_count': int(tweet['retweet_count']),
                        'favorite_count': int(tweet['favorite_count']),
                        'media_url': tweet['entities']['media'][0]['media_url'],
                        'retweeted': tweet['retweeted'],
                        'favorited': tweet['favorited']
                    })
            
            
        except Exception as e:
            print(e) 
            
        

In [10]:
# Criar dataframe de Tweets consultados na API
columns = tweets[0].keys()
df_tweets_api = pd.DataFrame(tweets, columns = columns)

------
<a id="avaliacao"></a>

# Avaliação

<a id="avaliacao:twitter-archive"></a>

## Avaliação: Twitter data archive

In [11]:
 df_twitter_arc.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1328  705970349788291072                    NaN                  NaN   
62    880095782870896641                    NaN                  NaN   
79    877556246731214848                    NaN                  NaN   
896   759099523532779520                    NaN                  NaN   
1438  696900204696625153                    NaN                  NaN   

                      timestamp  \
1328  2016-03-05 04:17:02 +0000   
62    2017-06-28 16:09:20 +0000   
79    2017-06-21 15:58:08 +0000   
896   2016-07-29 18:53:24 +0000   
1438  2016-02-09 03:35:31 +0000   

                                                 source  \
1328  <a href="http://twitter.com/download/iphone" r...   
62    <a href="http://twitter.com/download/iphone" r...   
79    <a href="http://twitter.com/download/iphone" r...   
896   <a href="http://twitter.com/download/iphone" r...   
1438  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1328  This is Lucy. She's a Venetian Kerploof. Suppo...                  NaN   
62    Please don't send in photos without dogs in th...                  NaN   
79    This is Penny. She's both pupset and fired pup...                  NaN   
896   Meet Toby. He has a drinking problem. Inflatab...                  NaN   
1438  This is Rosie. She's a Benebark Cumberpatch. S...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1328                       NaN                        NaN   
62                         NaN                        NaN   
79                         NaN                        NaN   
896                        NaN                        NaN   
1438                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1328  https://twitter.com/dog_rates/status/705970349...                12   
62    https://twitter.com/dog_rates/status/880095782...                11   
79    https://twitter.com/dog_rates/status/877556246...                12   
896   https://twitter.com/dog_rates/status/759099523...                 7   
1438  https://twitter.com/dog_rates/status/696900204...                12   

      rating_denominator   name doggo floofer pupper puppo  
1328                  10   Lucy  None    None   None  None  
62                    10   None  None    None   None  None  
79                    10  Penny  None    None   None  None  
896                   10   Toby  None    None   None  None  
1438                  10  Rosie  None    None   None  None

In [12]:
df_twitter_arc.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [13]:
df_twitter_arc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

<a id="twitter-arc:info"></a>
* Dados incompletos, contém apenas 2356 registros dos 5000 anunciados
* `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id` e `retweeted_status_user_id` em `float`
* `timestamp` e `retweeted_status_timestamp` em `string`

In [14]:
# Verificar se há IDs duplicados
df_twitter_arc[df_twitter_arc.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

<a id="twitter-arc:source"></a>

In [15]:
# Visualizar exemplos de dados da coluna `source`
df_twitter_arc.loc[:, 'source'].sample(10)

464     <a href="http://twitter.com/download/iphone" r...
980     <a href="http://twitter.com/download/iphone" r...
1385    <a href="http://twitter.com/download/iphone" r...
2029    <a href="http://twitter.com/download/iphone" r...
364     <a href="http://twitter.com/download/iphone" r...
420     <a href="http://twitter.com/download/iphone" r...
1633    <a href="http://twitter.com/download/iphone" r...
289     <a href="http://twitter.com/download/iphone" r...
1577    <a href="http://vine.co" rel="nofollow">Vine -...
1855    <a href="http://twitter.com/download/iphone" r...
Name: source, dtype: object

Variável `source` é uma URL para a Aplicação utilizada no tweet, sem valor para a unidade de observação

<a id="twitter-arc:name"></a>

In [16]:
# Visualizar exemplos da coluna `name`
df_twitter_arc.name.sample(10)

1471      None
702      Scout
1439      None
1413      None
1385      very
1507    Richie
217        Leo
2067      None
552      Rusty
1231     Lance
Name: name, dtype: object

In [17]:
df_twitter_arc[df_twitter_arc.name == 'None'].name.count()

745

* Nomes (`name`) preenchidos com valor literal igual a `'None'`, supostamente um valor não preenchido, vazio

<a id="twitter-arc:category"></id>

In [18]:
# Visualizar exemplo de dados das colunas 'doggo' a 'puppo'
df_twitter_arc.loc[:,'doggo':].sample(10)

doggo floofer  pupper puppo
278   None    None    None  None
1090  None    None    None  None
1967  None    None  pupper  None
679   None    None    None  None
2105  None    None    None  None
925   None    None    None  None
1908  None    None    None  None
2032  None    None    None  None
667   None    None    None  None
953   None    None    None  None

* Valores não preenchidos como 'None' literal

Uma variável em várias colunas

In [19]:
# Verificar padrão no preenchimento das classificações
df_twitter_arc.loc[:,'doggo':].nunique()

doggo      2
floofer    2
pupper     2
puppo      2
dtype: int64

In [20]:
df_twitter_arc.loc[:, 'doggo':'puppo'].isna().sum()

doggo      0
floofer    0
pupper     0
puppo      0
dtype: int64

<a id="twitter-arc:expanded_urls"></a>

Exemplos da coluna `expanded_urls`

In [21]:
df_twitter_arc.loc[:, 'expanded_urls'].sample(20)

1782    https://twitter.com/dog_rates/status/677687604...
1207    https://twitter.com/dog_rates/status/715733265...
1063    https://twitter.com/dog_rates/status/741067306...
381     https://twitter.com/dog_rates/status/827324948...
474     https://twitter.com/dog_rates/status/816091915...
259     https://twitter.com/dog_rates/status/843235543...
163     https://twitter.com/dog_rates/status/860276583...
2129    https://twitter.com/dog_rates/status/670290420...
1199    https://twitter.com/dog_rates/status/716791146...
2105    https://twitter.com/dog_rates/status/670474236...
2256    https://twitter.com/dog_rates/status/667766675...
923     https://twitter.com/dog_rates/status/755955933...
1545    https://twitter.com/dog_rates/status/689289219...
1639    https://twitter.com/dog_rates/status/684177701...
1403    https://twitter.com/dog_rates/status/699413908...
525     https://twitter.com/dog_rates/status/809084759...
1732    https://twitter.com/dog_rates/status/679828447...
2055    https:

In [22]:
df_twitter_arc.loc[:, 'expanded_urls'].isna().value_counts()

False    2297
True       59
Name: expanded_urls, dtype: int64

* Registros de `expanded_urls` com URLs para os twittes originais. Não apresenta valor a unidade em observacão

------
<a id="avaliacao:image-prediction"></a>

## Avaliação: Image prediction

In [23]:
df_prediction.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

<a id="image-prediction:info"></info>

In [24]:
df_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [25]:
df_prediction.sample(10)

tweet_id                                          jpg_url  \
1252  747963614829678593  https://pbs.twimg.com/media/CmFM7ngXEAEitfh.jpg   
1872  844979544864018432  https://pbs.twimg.com/media/C7n4aQ0VAAAohkL.jpg   
969   706516534877929472  https://pbs.twimg.com/media/Cc4NCQiXEAEx2eJ.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
313   671561002136281088  https://pbs.twimg.com/media/CVHdK-7WwAAsuyc.jpg   
1964  867072653475098625  https://pbs.twimg.com/media/DAElHfmUMAEH9lB.jpg   
1906  852311364735569921  https://pbs.twimg.com/media/C9QEqZ7XYAIR7fS.jpg   
97    667728196545200128  https://pbs.twimg.com/media/CUQ_QahUAAAVQjn.jpg   
1139  729463711119904772  https://pbs.twimg.com/media/Ch-TXpFXAAAwPGf.jpg   
502   675870721063669760  https://pbs.twimg.com/media/CWEs1b-WEAEhq82.jpg   

      img_num                p1   p1_conf  p1_dog                       p2  \
1252        1            kelpie  0.307672    True            Irish_terrier   
1872        3       tennis_ball  0.999281   False                   racket   
969         1  golden_retriever  0.772685    True       Labrador_retriever   
2072        1         Chihuahua  0.716012    True                 malamute   
313         1     Gordon_setter  0.469373    True  black-and-tan_coonhound   
1964        1  Blenheim_spaniel  0.352946    True                 papillon   
1906        1           barbell  0.971581   False                 dumbbell   
97          1            kuvasz  0.360159    True         golden_retriever   
1139        1   German_shepherd  0.829307    True                 Doberman   
502         1  golden_retriever  0.263892    True   Welsh_springer_spaniel   

       p2_conf  p2_dog                  p3       p3_conf  p3_dog  
1252  0.197486    True               dingo  1.054750e-01   False  
1872  0.000370   False   Shetland_sheepdog  1.320680e-04    True  
969   0.071665    True            golfcart  2.099310e-02   False  
2072  0.078253    True              kelpie  3.137890e-02    True  
313   0.270893    True          Rottweiler  1.532330e-01    True  
1964  0.211766    True            Pekinese  1.129520e-01    True  
1906  0.028418   False             go-kart  5.595040e-07   False  
97    0.293744    True  Labrador_retriever  2.706730e-01    True  
1139  0.022500    True             basenji  2.119010e-02    True  
502   0.184193    True              beagle  1.822410e-01    True

<a id="image-prediction:p-values"></a>

Valores não padronizados para as variáveis `p1`, `p2` e `p3`

In [26]:
# Visualização de exemplos de valores nas colunas `p1`, `p2` e `p3`
df_prediction.loc[:, ['p1', 'p2', 'p3']].sample(5)

p1                    p2           p3
1904              barrow  Bernese_mountain_dog  EntleBucher
1536        refrigerator              malinois       kelpie
1993            Cardigan              Pembroke       kelpie
281                goose             Chihuahua          hen
1076  miniature_pinscher     Italian_greyhound       beagle

<a id="image-prediction:duplicated-urls"></a>

Verificar a duplicidade de Tweet ID (`tweet_id`) e Imagens (`jpg_url`)

In [27]:
# Verificar a duplicidade de IDs
df_prediction.tweet_id.nunique()

2075

Verificar a duplicidade na coluna `jpg_url`

In [28]:
df_prediction.jpg_url.duplicated().value_counts()

False    2009
True       66
Name: jpg_url, dtype: int64

Há 66 imagens duplicadas

In [29]:
# Validar indicadores de confiabiliade da predição
df_prediction.loc[:, ['p1_conf', 'p2_conf', 'p3_conf']].max()

p1_conf    1.000000
p2_conf    0.488014
p3_conf    0.273419
dtype: float64

------
<a id="avaliacao:twitter-api"></a>

## Avaliação: Twitter API Requests

In [30]:
df_tweets_api.describe()

id  retweet_count  favorite_count
count  1.820000e+03    1820.000000     1820.000000
mean   7.237270e+17    2512.144505     6820.872527
std    5.777841e+16    4891.076169    11900.984651
min    6.660209e+17      11.000000        0.000000
25%    6.747671e+17     536.000000     1363.500000
50%    7.008223e+17    1118.000000     3153.000000
75%    7.617466e+17    2526.250000     7463.250000
max    8.924206e+17   82905.000000   163034.000000

<a id="tweets-api:info"></a>

In [31]:
df_tweets_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1820 entries, 0 to 1819
Data columns (total 10 columns):
id                       1820 non-null int64
created_at               1820 non-null object
in_reply_to_status_id    1820 non-null object
in_reply_to_user_id      1820 non-null object
is_quote_status          1820 non-null bool
retweet_count            1820 non-null int64
favorite_count           1820 non-null int64
media_url                1820 non-null object
retweeted                1820 non-null bool
favorited                1820 non-null bool
dtypes: bool(3), int64(3), object(4)
memory usage: 104.9+ KB


Os nomes das colunas `id` e `media_url` estão fora do padrão dos demais dataframes, respectivamente `tweet_id` e `jpg_url` 

In [32]:
df_tweets_api.sample(5)

id                      created_at  \
1096  686034024800862208  Sun Jan 10 03:57:12 +0000 2016   
906   701214700881756160  Sun Feb 21 01:19:47 +0000 2016   
59    852672615818899456  Thu Apr 13 23:59:28 +0000 2017   
540   750147208377409536  Tue Jul 05 02:00:06 +0000 2016   
1255  678255464182861824  Sat Dec 19 16:47:58 +0000 2015   

     in_reply_to_status_id in_reply_to_user_id  is_quote_status  \
1096                  None                None            False   
906                   None                None            False   
59                    None                None            False   
540                   None                None            False   
1255                  None                None            False   

      retweet_count  favorite_count  \
1096           1233            3238   
906            5396           12680   
59             2231           15191   
540            1028            3252   
1255            393            1637   

                                           media_url  retweeted  favorited  
1096  http://pbs.twimg.com/media/CYVIToGWQAAEZ_y.jpg      False      False  
906   http://pbs.twimg.com/media/Cbs3DOAXIAAp3Bd.jpg      False      False  
59    http://pbs.twimg.com/media/C9VNNp1XkAEWRFb.jpg      False      False  
540   http://pbs.twimg.com/media/CmkO57iXgAEOxX9.jpg      False      False  
1255  http://pbs.twimg.com/media/CWmlvxJU4AEAqaN.jpg      False      False

* `created_at` como `string`
* `in_reply_to_status_id` e `in_reply_to_userid` como `string`

<a id="tweets-api:duplicated"></a>
Verificar a duplicidade de valores

In [33]:
# Verificar se há IDs duplicados
df_tweets_api.id.duplicated().value_counts()

False    1820
Name: id, dtype: int64

In [34]:
# Verificar se há imagens duplicadas
df_tweets_api.media_url.duplicated().value_counts()

False    1759
True       61
Name: media_url, dtype: int64

Há tweets referenciando a mesma imagem

------

<a id="avaliacao:anotacoes"></a>

## Notas da avaliação
### Qualidade

#### `df_twitter_arc` - Twitter archive
1. [Base incompleta, contém apenas **2356** registros dos **5000** anunciados](#twitter-arc:info)
* [Coluna `source` sem valor à unidade de obsevação](#twitter-arc:source)
* [Nomes (`name`) preenchidos com valor literal igual a `'None'`, vazio](#twitter-arc:name)
* [Coluna `expanded_urls` sem valor à unidade de observação](#twitter-arc:expanded_urls)
* [Categorias \[`doggo`, `floofer`, `pupper`, `puppo`\] com literal `'None'`](#twitter-arc:category)


#### `df_prediction` - Image prediction
1. [Previsões \[`p1`, `p2`, `p3`\] com nomes não padronizados](#image-prediction:p-values)
* [Duplicidade na coluna `jgp_url`](#image-prediction:duplicated-urls)


#### `df_tweets_api` - Twitter API
1. [Diferentes Tweets referenciando a mesma imagem, duplicidade](#tweets-api:duplicated)


### Organização

#### `df_twitter_arc` - Twitter archive
1. [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#twitter-arc:info)
* [As colunas `timestamp` e `retweet_status_timestamp` estão em `string`](#twitter-arc:info)
* [Uma variável em várias colunas, `doggo`, `floofer`, `pupper` e `puppo`](#twitter-arc:category)


#### `df_prediction` - Image prediction
1. [Colunas como variáveis na identificação das predições](#image-prediction:info)


#### `df_tweets_api` - Twitter API
1. [Os nomes das colunas `id` e `media_url` estão fora do padrão dos demais dataframes, respectivamente `tweet_id` e `jpg_url` ](#tweets-api:info)
* [Data de criação (`created_at`) como `string`](#tweets-api:info)
* [As colunas `in_reply_to_status_id` e `in_reply_to_user_id` em `string`](#tweets-api:info)

------

<a id="limpeza"></a>

# Limpeza

<a id="limpeza:definicao"></a>

## Limpeza: Definição 

#### `df_twitter_arc` - Twitter archive
1. [Base incompleta, de **5000**, apenas **2356** estão disponíveis](#limpeza:twitter-arc:completeness)
    * OK - Nada a ser feito
* [Coluna `source` sem valor à unidade de observação](#limpeza:twitter-arc:source)
    * O conteúdo apresenta um link para download do App Twitter, portanto não agrega valor à unidade de observação, a coluna pode ser removida
* [Nomes (`name`) não preenchidos com valor literal igual a `'None'`](#limpeza:twitter-arc:name)
    * Alterar valores como 'None' para `np.nan`
* [Coluna `expanded_urls` sem valor à unidade de observação](#limpeza:twitter-arc:expanded_urls)(#twitter-arc:expanded_urls)
    * Não apresentam valor à unidade de observação, são URLs para os tweets em questão. Remover a coluna `expanded_urls`.
* [Categorias \[doggo, floofer, pupper, puppo\] com literal `'None'`](#limpeza:twitter-arc:category)
    * Aplicar valor `np.nan` para nomes iguais a 'None'   
* [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#limpeza:twitter-arc:info)
    * A coluna, `retweeted_status_id` ou `retweeted_status_user_id`, podem ser convertidas para `Boolean` como auxílio a identificação de tweets originais
    * Após a criação de uma coluna indicando se o tweet é um retweet, as colunas de retweet, `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`, podem ser descartadas
    * As colunas `in_reply_to_status_id` e `in_reply_to_user_id` serão retratadas no dataframe `df_tweets_api`
* [As colunas `timestamp` e `retweet_status_timestamp` estão em `string` e deveriam ser do tipo `datetime`](#limpeza:twitter-arc:datetime)
    * Converter valores para `datetime`
* [Uma variável em várias colunas, `doggo`, `floofer`, `pupper` e `puppo`](#limpeza:twitter-arc:category-merge)
    * Reduzir a variável para uma coluna, `category`


#### `df_prediction` - Image prediction
1. [Previsões \[`p1`, `p2`, `p3`\] com nomes não padronizados](#limpeza:image-prediction:names)
    * Converter os valores para 'lower case' e substituir '_' por espaços
* [Previsões repetidas sobre a mesma imagem](#limpeza:image-prediction:duplicated-urls)
    * Descartar predições sobre imagens repetidas
* [Colunas como variáveis na identificação das predições](#limpeza:image-prediction:info)
    * Transpor as variáveis nas colunas para linhas


#### `df_tweets_api` - Twitter API
1. [Os nomes das colunas `id` e `media_url` estão fora do padrão dos demais dataframes, respectivamente `tweet_id` e `jpg_url` ](#limpeza:tweets-api:columns)
* [Diferentes Tweets referenciando a mesma imagem, duplicidade](#limpeza:tweets-api:duplicated)
* [Data de criação (`created_at`) como `string`](#limpeza:tweets-api:datetime)
    * Converter datas em string para o formato `datetime`
* [Valor do id na coluna `in_reply_to_status_id` e `in_reply_to_user_id` como `string`](#limpeza:tweets-api:info)
    * Converter para Inteiro as colunas `in_reply_to_status_id` e `in_reply_to_user_id` e atribuir np.nan as que estiverem sem valor

<a id="limpeza:twitter-archive"></a>

## Limpeza: Twitter archive

In [35]:
# Criar cópia do dataframe para a limpeza
df_arch_clean = df_twitter_arc.copy()

<a id="limpeza:twitter-arc:completeness"></a>

### 1. [Base incompleta, de 5000, apenas 2356 estão disponíveis](#limpeza:twitter-arc:completeness)
Nada a ser feito

In [36]:
df_arch_clean.shape

(2356, 17)

<a id="limpeza:twitter-arc:source"></a>

### 2. [Coluna `source` sem valor à unidade de observação](#limpeza:twitter-arc:source)
* O conteúdo apresenta um link para download do App Twitter, portanto não agrega valor à unidade de observação, a coluna pode ser removida

In [37]:
# Remoção da coluna `source`
df_arch_clean.drop('source', axis='columns', inplace=True)

#### Teste: Remoção coluna `source`

In [38]:
# Conferir que a coluna `source` foi removida
df_arch_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

<a id="limpeza:twitter-arc:name"></a>

### 3. [Nomes (`name`) não preenchidos com valor literal igual a `'None'`](#limpeza:twitter-arc:name)
    * Alterar valores como 'None' para `np.nan`

In [39]:
# Contagem de nomes iguais a 'None'
df_arch_clean[df_arch_clean.name.str.lower() == 'none'].name.count()

745

In [40]:
# Função para substituição de valores literais 'None' por np.nan
def set_nan_at_none(value):
    if str(value).lower() == 'none':
        return np.nan
    else:
        return value

In [41]:
# Aplicar `np.nan` aos nomes iguais a 'None'
df_arch_clean.name = df_arch_clean.name.apply(set_nan_at_none)

#### Teste: Aplicação de `np.nan` aos nomes como 'None'

In [42]:
df_arch_clean.name.isnull().sum()

745

In [43]:
df_arch_clean[df_arch_clean.name.str.lower() == 'none'].name.count()

0

In [44]:
df_arch_clean.name.sample(10)

178         Oscar
829           NaN
1659          NaN
1717          NaN
1289    Jebberson
211        Astrid
1842          NaN
216          Kyle
946           NaN
1492          NaN
Name: name, dtype: object

<a id="limpeza:twitter-arc:expanded_urls"></a>

### 4. [Coluna `expanded_urls` sem valor à unidade de observação](#limpeza:twitter-arc:expanded_urls)
* Não apresentam valor à unidade de observação, são URLs para os tweets em questão. Remover a coluna `expanded_urls`

In [45]:
df_arch_clean.drop('expanded_urls', axis='columns', inplace=True)

#### Teste: Remoção da coluna `expanded_urls`

In [46]:
df_arch_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'rating_numerator', 'rating_denominator',
       'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

<a id="limpeza:twitter-arc:category"></a>

### 5. [Categorias \[`doggo`, `floofer`, `pupper`, `puppo`\] com literal `'None'`](#limpeza:twitter-arc:category)
* Aplicar valor `np.nan` para nomes iguais a 'None'   

In [47]:
# Trocar valores 'None' para np.nan nas colunas `doggo`, `floofer`, `pupper`, `puppo`
df_arch_clean.doggo = df_arch_clean.doggo.apply(set_nan_at_none)
df_arch_clean.doggo.isna().value_counts()

True     2259
False      97
Name: doggo, dtype: int64

In [48]:
df_arch_clean.floofer = df_arch_clean.floofer.apply(set_nan_at_none)
df_arch_clean.floofer.isna().value_counts()

True     2346
False      10
Name: floofer, dtype: int64

In [49]:
df_arch_clean.pupper = df_arch_clean.pupper.apply(set_nan_at_none)
df_arch_clean.pupper.isna().value_counts()

True     2099
False     257
Name: pupper, dtype: int64

In [50]:
df_arch_clean.puppo = df_arch_clean.puppo.apply(set_nan_at_none)
df_arch_clean.puppo.isna().value_counts()

True     2326
False      30
Name: puppo, dtype: int64

In [51]:
# Soma de valores preenchidos nas classificações
sum_categories = 0
sum_categories += df_arch_clean.doggo.notna().sum()
sum_categories += df_arch_clean.floofer.notna().sum()
sum_categories += df_arch_clean.pupper.notna().sum()
sum_categories += df_arch_clean.puppo.notna().sum()

sum_categories

394

Apenas 394 dos 2346 registros foram classificados

<a id="limpeza:twitter-arc:info"></a>

### 6. [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#limpeza:twitter-arc:info)
* Linhas com algum valor para as colunas `retweeted_status_id` ou `retweeted_status_user_id` podem ser descartadas por indicarem retweet, e o objetivo é analisar somente tweets originais
* Após a remoção retweets, as colunas `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`, podem ser descartadas
* As colunas `in_reply_to_status_id` e `in_reply_to_user_id` serão retratadas no dataframe `df_tweets_api`

#### Identificação dos tweets originais
df_arch_clean.loc[:,'retweeted_status_id'].isna().value_counts()

In [52]:
df_arch_clean.retweeted_status_id.size

2356

In [53]:
df_arch_clean.loc[:,'retweeted_status_id'].isna().value_counts()

True     2175
False     181
Name: retweeted_status_id, dtype: int64

Dos 2356 registros, 181 são retweets

In [54]:
# Remoção dos retweets
df_arch_clean.drop(df_arch_clean[df_arch_clean.retweeted_status_id.notna()].index, axis=0, inplace=True)

#### Teste: Remoção de Retweets

In [55]:
# Consultar a existência de colunas com retweet_status_preenchido
df_arch_clean.loc[:,'retweeted_status_id'].isna().value_counts()

True    2175
Name: retweeted_status_id, dtype: int64

Remoção das colunas referentes a retweets

In [56]:
df_arch_clean.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis='columns', inplace=True)

#### Teste: Verificação da remoção das colunas referentes a retweets

In [57]:
df_arch_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'text', 'rating_numerator', 'rating_denominator', 'name', 'doggo',
       'floofer', 'pupper', 'puppo'],
      dtype='object')

<a id="limpeza:twitter-arc:datetime"></a>

### 7. [As colunas `timestamp` e `retweet_status_timestamp` estão em `string` e deveriam ser do tipo `datetime`](#limpeza:twitter-arc:datetime)
* Converter valores para `datetime`

Haja vista que a coluna `retweet_status_timestamp` foi removida no passo anterior, não será necessária a sua conversão

In [58]:
df_arch_clean.loc[:, ['timestamp']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 1 columns):
timestamp    2175 non-null object
dtypes: object(1)
memory usage: 34.0+ KB


In [59]:
df_arch_clean.timestamp = pd.to_datetime(df_arch_clean.timestamp)

#### Teste: Veriificação da conversão de `string` para `datetime`

In [60]:
df_arch_clean.loc[:, ['timestamp']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 1 columns):
timestamp    2175 non-null datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1)
memory usage: 34.0 KB


<a id="limpeza:twitter-arc:category-merge"></a>

### 8. [Uma variável em várias colunas, `doggo`, `floofer`, `pupper` e `puppo`](#limpeza:twitter-arc:category-merge)
* Reduzir a variável para uma coluna, `category`

In [61]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null datetime64[ns, UTC]
text                     2175 non-null object
rating_numerator         2175 non-null int64
rating_denominator       2175 non-null int64
name                     1495 non-null object
doggo                    87 non-null object
floofer                  10 non-null object
pupper                   234 non-null object
puppo                    25 non-null object
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(6)
memory usage: 220.9+ KB


In [62]:
df_arch_clean.loc[:, 'doggo':].notna().sum()

doggo       87
floofer     10
pupper     234
puppo       25
dtype: int64

In [63]:
df_arch_clean.loc[:, 'doggo':].notna().sum().sum()

356

De **2175** registros, apenas **356** estão classificados

Criar uma lista das categorias para mesclar as colunas, considerando a coluna a esquerda como prioritaria sobre valores nulos a direita.

In [64]:
category = df_arch_clean.doggo
category = category.combine_first(df_arch_clean.floofer)
category = category.combine_first(df_arch_clean.pupper)
category = category.combine_first(df_arch_clean.puppo)

df_arch_clean['category'] = category

In [65]:
df_arch_clean.category.notna().sum()

344

Após a unificação das colunas de classificação, pode-se ver que resultou em apenas **344** registros, diferente dos **356** iniciais. O que indica que haviam registros com mais de uma classificação.

In [66]:
df_arch_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis='columns', inplace=True)

#### Teste: Verificar a criação da coluna `category` a remoção das colunas `doggo`, `floofer`, `pupper` e `puppo`. 

In [67]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null datetime64[ns, UTC]
text                     2175 non-null object
rating_numerator         2175 non-null int64
rating_denominator       2175 non-null int64
name                     1495 non-null object
category                 344 non-null object
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(3)
memory usage: 169.9+ KB


------
<a id="limpeza:image-prediction"></a>

## Limpeza: Image prediction 

In [68]:
# Cópia do dataframe para limpeza
df_predict_clean = df_prediction.copy()

<a id="limpeza:image-prediction:names"></a>

### 1. [Previsões \[`p1`, `p2`, `p3`\] com nomes não padronizados](#limpeza:image-prediction:names)
* Converter os valores para 'lower case' e substituir '_' por espaços

In [69]:
clean_prediction_name = lambda value: str(value).lower().replace('_', ' ')

In [70]:
df_predict_clean.loc[:, ['p1', 'p2', 'p3']].sample(5)

p1                  p2              p3
2027    golden_retriever  Labrador_retriever  cocker_spaniel
1221  Labrador_retriever            ice_bear         whippet
1618    golden_retriever  Labrador_retriever  cocker_spaniel
763                crane                pier       bell_cote
585           Eskimo_dog      Siberian_husky       Chihuahua

Aplicar a padronização dos nomes em cada uma das colunas de predição

In [71]:
df_predict_clean.p1 = df_predict_clean.p1.apply(clean_prediction_name)

In [72]:
df_predict_clean.p2 = df_predict_clean.p2.apply(clean_prediction_name)

In [73]:
df_predict_clean.p3 = df_predict_clean.p3.apply(clean_prediction_name)

#### Teste: Verificar padronização dos nomes das predições

In [74]:
df_predict_clean.loc[:, ['p1', 'p2', 'p3']].sample(5)

p1                p2              p3
284       guinea pig       wood rabbit          borzoi
621           kuvasz           samoyed  great pyrenees
1377        malamute              jean        keeshond
600   french bulldog         chihuahua        cardigan
1766  great pyrenees  golden retriever        pekinese

<a id="limpeza:image-prediction:duplicated-urls"></a>

### 2. [Predições repetidas sobre a mesma imagem](#limpeza:image-prediction:duplicated-urls)
* Descartar predições sobre imagens repetidas

In [75]:
df_predict_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [76]:
df_predict_clean.loc[:, ['jpg_url', 'p1', 'p2', 'p3']].duplicated().value_counts()

False    2009
True       66
dtype: int64

Dos **2075** registros, **66** são sobre a mesma imagem. Entretanto, seguindo o objetivo de trabalhar com apenas Tweets originais, além de descartar duplicidade é interessante também descartar as refêrencias a Retweets. Para isso, será feita uma consulta a seleção de Tweets originais, que foi efetuada no passo [6 da limpeza do Twitter Archive](#limpeza:twitter-arc:info)

In [77]:
df_predict_clean = df_predict_clean.loc[df_predict_clean.tweet_id.isin(df_arch_clean.tweet_id)]

#### Teste: Seleção de Tweets originais

In [78]:
df_predict_clean.tweet_id.isin(df_arch_clean.tweet_id).value_counts()

True    1994
Name: tweet_id, dtype: int64

Apenas dos **2075**, **1994** são de Tweets originais

------

Dos Tweets originai ainda é necessário verificar a existência de duplicidade de predições.

In [79]:
df_predict_clean.loc[:, ['jpg_url']].duplicated().value_counts()

False    1994
dtype: int64

Nenhuma duplicidade foi encontrada após a remoção das referências de Retweets

In [80]:
df_predict_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    1994 non-null int64
jpg_url     1994 non-null object
img_num     1994 non-null int64
p1          1994 non-null object
p1_conf     1994 non-null float64
p1_dog      1994 non-null bool
p2          1994 non-null object
p2_conf     1994 non-null float64
p2_dog      1994 non-null bool
p3          1994 non-null object
p3_conf     1994 non-null float64
p3_dog      1994 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 161.6+ KB


<a id="limpeza:image-prediction:info"></a>

### 3. [Colunas como variáveis na identificação das predições](#limpeza:image-prediction:info)
* 3.1. Transpor as variáveis nas colunas para linhas:
    * 3.1.1. [`[p1, p2, p3]`, para `prediction_try`](#limpeza:image-prediction:try)
    * 3.1.2. [`[p1_conf, p2_conf, p3_conf]`, para `confidence`](#limpeza:image-prediction:conf)
    * 3.1.3. [`[p1_dog, p2_dog, p3_dog]`, para `is_a_dog`](#limpeza:image-prediction:dog)
* 3.2. [Unir os dataframes de prediction](#limpeza:image-prediction:merge)

In [81]:
df_predict_clean.sample(5)

tweet_id                                            jpg_url  \
985   707610948723478529    https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1316  755110668769038337  https://pbs.twimg.com/ext_tw_video_thumb/75511...   
2024  881666595344535552    https://pbs.twimg.com/media/DDxPFwbWAAEbVVR.jpg   
1139  729463711119904772    https://pbs.twimg.com/media/Ch-TXpFXAAAwPGf.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   

      img_num                    p1   p1_conf  p1_dog                p2  \
985         1      golden retriever  0.383223    True    cocker spaniel   
1316        1    labrador retriever  0.708974    True  golden retriever   
2024        1                saluki  0.529012    True      afghan hound   
1139        1       german shepherd  0.829307    True          doberman   
5           1  bernese mountain dog  0.651137    True  english springer   

       p2_conf  p2_dog                          p3   p3_conf  p3_dog  
985   0.165930    True    chesapeake bay retriever  0.118199    True  
1316  0.114314    True              great pyrenees  0.065813    True  
2024  0.250003    True            golden retriever  0.160739    True  
1139  0.022500    True                     basenji  0.021190    True  
5     0.263788    True  greater swiss mountain dog  0.016199    True

<a id="limpeza:image-prediction:try"></a>

#### 3.1.1. [`[p1, p2, p3]`, para `prediction_try`](#limpeza:image-prediction:try)

In [82]:
# Organizando as variáveis de tentativa de predição e seu palpite
df_predict_try = pd.melt(df_predict_clean,
        id_vars=['tweet_id', 'jpg_url'],
        value_vars=['p1', 'p2', 'p3'],
        var_name=['prediction_try'],
        value_name='prediction'
       )

In [83]:
df_predict_try.sample(5)

tweet_id                                            jpg_url  \
1316  756303284449767430    https://pbs.twimg.com/media/Cn7tyyZWYAAPlAY.jpg   
5529  795076730285391872    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   
1857  859074603037188101    https://pbs.twimg.com/media/C-wLyufW0AA546I.jpg   
1331  758467244762497024  https://pbs.twimg.com/ext_tw_video_thumb/75846...   
2509  676440007570247681    https://pbs.twimg.com/media/CWMyl9EWUAAnZJ0.jpg   

     prediction_try          prediction  
1316             p1    golden retriever  
5529             p3               swing  
1857             p1            revolver  
1331             p1  labrador retriever  
2509             p2          toy poodle

Conversão dos valores da tentativa para número inteiro

In [84]:
# Metodo para retirar o índice da tentativa dos nomes das colunas
get_prediction_index = lambda value: int(re.sub(r'[^\d]', '', str(value)))

In [85]:
df_predict_try.prediction_try = df_predict_try.prediction_try.apply(get_prediction_index)

In [86]:
df_predict_try.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5982 entries, 0 to 5981
Data columns (total 4 columns):
tweet_id          5982 non-null int64
jpg_url           5982 non-null object
prediction_try    5982 non-null int64
prediction        5982 non-null object
dtypes: int64(2), object(2)
memory usage: 187.0+ KB


In [87]:
df_predict_try.sample(5)

tweet_id                                          jpg_url  \
491   675707330206547968  https://pbs.twimg.com/media/CWCYOqWUAAARmGr.jpg   
4553  678341075375947776  https://pbs.twimg.com/media/CWnznDTU4AAa-6P.jpg   
3856  860184849394610176  https://pbs.twimg.com/media/C-_9jWWUwAAnwkd.jpg   
2190  669625907762618368  https://pbs.twimg.com/media/CUr9NjgU8AEpf5w.jpg   
1131  728653952833728512  https://pbs.twimg.com/media/Chyy5lQWUAEzxSL.jpg   

      prediction_try          prediction  
491                1          bath towel  
4553               3  labrador retriever  
3856               2             gorilla  
2190               2    golden retriever  
1131               1        window shade

<a id="limpeza:image-prediction:conf"></a>

#### 3.1.2. [`[p1_conf, p2_conf, p3_conf]`, para `confidence`](#limpeza:image-prediction:conf)

In [88]:
# Movendo o nível de confiança das predições para uma coluna apenas
df_predict_conf = pd.melt(df_predict_clean,
                         id_vars=['tweet_id', 'jpg_url'],
                         value_vars=['p1_conf', 'p2_conf', 'p3_conf'],
                         var_name=['prediction_try'],
                         value_name='confidence'
                        )

In [89]:
df_predict_conf.sample(5)

tweet_id                                            jpg_url  \
1362  762471784394268675  https://pbs.twimg.com/ext_tw_video_thumb/76247...   
543   677314812125323265    https://pbs.twimg.com/media/CWZOOIUW4AAQrX_.jpg   
5367  766793450729734144    https://pbs.twimg.com/media/CqQykxrWYAAlD8g.jpg   
2660  683030066213818368    https://pbs.twimg.com/media/CXqcOHCUQAAugTB.jpg   
2828  694206574471057408    https://pbs.twimg.com/media/CaJRMPQWIAA1zL9.jpg   

     prediction_try  confidence  
1362        p1_conf    0.540276  
543         p1_conf    0.924127  
5367        p3_conf    0.072699  
2660        p2_conf    0.193804  
2828        p2_conf    0.155720

In [90]:
# Atribuir apenas valores inteiros aos indicadores de tentativas
df_predict_conf.prediction_try = df_predict_conf.prediction_try.apply(get_prediction_index)

In [91]:
df_predict_conf.sample(5)

tweet_id                                          jpg_url  \
804   691820333922455552  https://pbs.twimg.com/media/CZnW7JGW0AA83mn.jpg   
4264  670995969505435648  https://pbs.twimg.com/media/CU_bRIEWcAAUVC7.jpg   
638   681302363064414209  https://pbs.twimg.com/media/CXR44l9WcAAcG_N.jpg   
3615  813096984823349248  https://pbs.twimg.com/media/C0izZULWgAAKD-F.jpg   
3373  766793450729734144  https://pbs.twimg.com/media/CqQykxrWYAAlD8g.jpg   

      prediction_try  confidence  
804                1    0.332756  
4264               3    0.024285  
638                1    0.326259  
3615               2    0.117003  
3373               2    0.197513

In [92]:
df_predict_conf.shape

(5982, 4)

<a id="limpeza:image-prediction:dog"></a>

#### 3.1.3. [`[p1_dog, p2_dog, p3_dog]`, para `is_a_dog`](#limpeza:image-prediction:dog)

In [93]:
# Movendo o resultado da predição, `p#_dog` para `is_a_dog`
df_predict_dog = pd.melt(df_predict_clean,
                        id_vars=['tweet_id', 'jpg_url'],
                        value_vars=['p1_dog', 'p2_dog', 'p3_dog'],
                        var_name=['prediction_try'],
                        value_name='is_a_dog')

In [94]:
df_predict_dog.sample(5)

tweet_id                                          jpg_url  \
1983  890006608113172480  https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg   
4336  672488522314567680  https://pbs.twimg.com/media/CVUovvHWwAAD-nu.jpg   
860   697255105972801536  https://pbs.twimg.com/media/Ca0lzzmWwAA5u56.jpg   
482   675483430902214656  https://pbs.twimg.com/media/CV_MmGZU8AAggM6.jpg   
1853  857989990357356544  https://pbs.twimg.com/media/C-gxV9ZXkAIBL-S.jpg   

     prediction_try  is_a_dog  
1983         p1_dog      True  
4336         p3_dog      True  
860          p1_dog      True  
482          p1_dog     False  
1853         p1_dog      True

In [95]:
# Atribuir apenas valores inteiros aos indicadores de tentativas
df_predict_dog.prediction_try = df_predict_dog.prediction_try.apply(get_prediction_index)

In [96]:
df_predict_dog.sample(5)

tweet_id                                          jpg_url  \
4486  675845657354215424  https://pbs.twimg.com/media/CWEWClfW4AAnqhG.jpg   
2088  667724302356258817  https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   
4900  700890391244103680  https://pbs.twimg.com/media/CboQFolWIAE04qE.jpg   
2817  693262851218264065  https://pbs.twimg.com/media/CZ724fDUYAAytS-.jpg   
5474  786233965241827333  https://pbs.twimg.com/media/CulDnZpWcAAGbZ-.jpg   

      prediction_try  is_a_dog  
4486               3      True  
2088               2     False  
4900               3      True  
2817               2      True  
5474               3      True

In [97]:
df_predict_dog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5982 entries, 0 to 5981
Data columns (total 4 columns):
tweet_id          5982 non-null int64
jpg_url           5982 non-null object
prediction_try    5982 non-null int64
is_a_dog          5982 non-null bool
dtypes: bool(1), int64(2), object(1)
memory usage: 146.1+ KB


<a id="limpeza:image-prediction:merge"></a>

#### 3.2. [Unir os dataframes de prediction](#limpeza:image-prediction:merge)

In [98]:
# Merge
df_predict_clean = pd.merge(df_predict_try, df_predict_conf, how='left', on=['tweet_id', 'jpg_url', 'prediction_try'])

In [99]:
df_predict_clean.shape

(5982, 5)

In [100]:
df_predict_clean.sample(5)

tweet_id                                          jpg_url  \
4777  690690673629138944  https://pbs.twimg.com/media/CZXTgKkWwAA5UZJ.jpg   
2205  670040295598354432  https://pbs.twimg.com/media/CUx2F6lVEAAvFev.jpg   
1397  770655142660169732  https://pbs.twimg.com/media/CrHqwjWXgAAgJSe.jpg   
1708  827600520311402496  https://pbs.twimg.com/media/C3w6RYbWQAAEQ25.jpg   
618   680473011644985345  https://pbs.twimg.com/media/CXGGlzvWYAArPfk.jpg   

      prediction_try        prediction  confidence  
4777               3       maltese dog    0.089625  
2205               2            borzoi    0.026660  
1397               1    madagascar cat    0.494803  
1708               1          pembroke    0.325638  
618                1  lakeland terrier    0.796694

In [101]:
df_predict_clean = pd.merge(df_predict_clean, df_predict_dog, how='left', on=['tweet_id', 'jpg_url', 'prediction_try'])

In [102]:
df_predict_clean.shape

(5982, 6)

**Teste: Verificar Merge dos dataframes de predição**

In [103]:
df_predict_clean.sample(5)

tweet_id                                            jpg_url  \
1571  801538201127157760    https://pbs.twimg.com/media/Cx-itFWWIAAZu7l.jpg   
153   668815180734689280    https://pbs.twimg.com/media/CUgb21RXIAAlff7.jpg   
5551  800141422401830912    https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
906   700505138482569216    https://pbs.twimg.com/media/Cbixs3vUUAAqHHN.jpg   
569   678399652199309312  https://pbs.twimg.com/ext_tw_video_thumb/67839...   

      prediction_try          prediction  confidence  is_a_dog  
1571               1            pembroke    0.550506      True  
153                1             redbone    0.461172      True  
5551               3  labrador retriever    0.022977      True  
906                1          bath towel    0.449684     False  
569                1               swing    0.929196     False

------
<a id="limpeza:twitter-api"></a>

## Limpeza: Twitter API

In [104]:
# Cópia do dataframe para limpeza
df_api_clean = df_tweets_api.copy()

In [105]:

df_api_clean.sample(3)

id                      created_at  \
1146  683391852557561860  Sat Jan 02 20:58:09 +0000 2016   
1788  666421158376562688  Tue Nov 17 01:02:40 +0000 2015   
880   704054845121142784  Sun Feb 28 21:25:30 +0000 2016   

     in_reply_to_status_id in_reply_to_user_id  is_quote_status  \
1146                  None                None            False   
1788                  None                None            False   
880                   None                None            False   

      retweet_count  favorite_count  \
1146           2522            7925   
1788            111             311   
880             956            3032   

                                           media_url  retweeted  favorited  
1146  http://pbs.twimg.com/media/CXvlQ2zW8AAE0tp.jpg      False      False  
1788  http://pbs.twimg.com/media/CT-aggCXAAIMfT3.jpg      False      False  
880   http://pbs.twimg.com/media/CcVOJEcXEAM0FHL.jpg      False      False

<a id="limpeza:tweets-api:columns"></a>

### 1. [Os nomes das colunas `id` e `media_url` estão fora do padrão dos demais dataframes, respectivamente `tweet_id` e `jpg_url` ](#limpeza:tweets-api:columns)

In [106]:
df_api_clean.rename(columns={'id':'tweet_id', 'media_url': 'jpg_url'}, inplace=True)

**Teste: Verificar colunas renomeadas**

In [107]:
df_api_clean.columns

Index(['tweet_id', 'created_at', 'in_reply_to_status_id',
       'in_reply_to_user_id', 'is_quote_status', 'retweet_count',
       'favorite_count', 'jpg_url', 'retweeted', 'favorited'],
      dtype='object')

<a id="limpeza:tweets-api:duplicated"></a>

### 2. [Diferentes Tweets referenciando a mesma imagem, duplicidade](#limpeza:tweets-api:duplicated)

Assim como na [limpeza de duplicidade de imagens no dataframe de predições](#limpeza:image-prediction:duplicated-urls), é importante de remover os tweets não originais, os retweets. Com auxílio do dataframe `df_arch_clean`, que possui somente referências a tweets originais, será possível identifica-los e descartar os tweets indesejados.

In [108]:
df_api_clean.shape

(1820, 10)

In [109]:
df_api_clean = df_api_clean.loc[df_api_clean.tweet_id.isin(df_arch_clean.tweet_id)]

In [110]:
df_api_clean.shape

(1746, 10)

Ao remover os ids que referenciavam a **retweets**, houve uma redução de **74** registros, de **1820** para **1746**

<a id="limpeza:tweets-api:datetime"></a>

### 3. [Data de criação (`created_at`) como `string`](#limpeza:tweets-api:datetime)
* Converter datas em string para o formato `datetime`

In [111]:
df_api_clean.created_at = pd.to_datetime(df_api_clean.created_at)

**Teste: Validar tipo de variável em `created_at`**

In [112]:
df_api_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1746 entries, 0 to 1819
Data columns (total 10 columns):
tweet_id                 1746 non-null int64
created_at               1746 non-null datetime64[ns, UTC]
in_reply_to_status_id    1746 non-null object
in_reply_to_user_id      1746 non-null object
is_quote_status          1746 non-null bool
retweet_count            1746 non-null int64
favorite_count           1746 non-null int64
jpg_url                  1746 non-null object
retweeted                1746 non-null bool
favorited                1746 non-null bool
dtypes: bool(3), datetime64[ns, UTC](1), int64(3), object(3)
memory usage: 114.2+ KB


In [113]:
df_api_clean.sample(5)

tweet_id                created_at in_reply_to_status_id  \
1651  668988183816871936 2015-11-24 03:03:06+00:00                  None   
741   718234618122661888 2016-04-08 00:30:51+00:00                  None   
1230  679132435750195208 2015-12-22 02:52:45+00:00                  None   
509   754747087846248448 2016-07-17 18:38:22+00:00                  None   
1586  670420569653809152 2015-11-28 01:54:54+00:00                  None   

     in_reply_to_user_id  is_quote_status  retweet_count  favorite_count  \
1651                None            False            479             901   
741                 None            False           1058            4008   
1230                None            False           1225            3077   
509                 None            False            555            2700   
1586                None            False            321             630   

                                             jpg_url  retweeted  favorited  
1651  http://pbs.twimg.com/media/CUi5M7TXIAAY0gj.jpg      False      False  
741   http://pbs.twimg.com/media/CfeukpmW4AEGjOE.jpg      False      False  
1230  http://pbs.twimg.com/media/CWzDWOkXAAAP0k7.jpg      False      False  
509   http://pbs.twimg.com/media/CnlmeL3WgAA4c84.jpg      False      False  
1586  http://pbs.twimg.com/media/CU3P82RWEAAIVrE.jpg      False      False

<a id="limpeza:tweets-api:info"></a>

### 4. [Valor do id na coluna `in_reply_to_status_id` e `in_reply_to_user_id` como `string`](#limpeza:tweets-api:info)
* Converter para Inteiro as colunas `in_reply_to_status_id` e `in_reply_to_user_id` e atribuir np.nan as que estiverem sem valor

In [114]:
b = df_api_clean.copy()

In [115]:
df_api_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1746 entries, 0 to 1819
Data columns (total 10 columns):
tweet_id                 1746 non-null int64
created_at               1746 non-null datetime64[ns, UTC]
in_reply_to_status_id    1746 non-null object
in_reply_to_user_id      1746 non-null object
is_quote_status          1746 non-null bool
retweet_count            1746 non-null int64
favorite_count           1746 non-null int64
jpg_url                  1746 non-null object
retweeted                1746 non-null bool
favorited                1746 non-null bool
dtypes: bool(3), datetime64[ns, UTC](1), int64(3), object(3)
memory usage: 114.2+ KB


Converter os valores em `None` para np.nan em `in_reply_to_status_id` e `in_reply_to_user_id`

In [116]:
def set_zero_at_none(value):
    if str(value).lower() == 'none':
        return '0'
    else:
        return value

In [117]:
df_api_clean.in_reply_to_status_id = df_api_clean.in_reply_to_status_id.apply(set_zero_at_none)

In [118]:
df_api_clean.in_reply_to_user_id = df_api_clean.in_reply_to_user_id.apply(set_zero_at_none)

In [119]:
df_api_clean.loc[:, ['in_reply_to_user_id', 'in_reply_to_status_id']] = df_api_clean.loc[:, ['in_reply_to_user_id', 'in_reply_to_status_id']].astype('int64')

Listar exemplos com valores 'None'substituídos por '0' e convertidos para `int64`

In [120]:
df_api_clean.loc[:, ['in_reply_to_user_id', 'in_reply_to_status_id']].sample(5)

in_reply_to_user_id  in_reply_to_status_id
275                     0                      0
394                     0                      0
475                     0                      0
903                     0                      0
1594                    0                      0

Converter valores iguais a 0 por `np.nan`

In [121]:
set_nan_at_zero = lambda value: np.nan if value is 0 else value

In [122]:
df_api_clean.in_reply_to_status_id = df_api_clean.in_reply_to_status_id.apply(set_nan_at_zero)

In [123]:
df_api_clean.in_reply_to_user_id = df_api_clean.in_reply_to_user_id.apply(set_nan_at_zero)

**Teste: Confirmar conversão para inteiro das colunas `in_reply_to_status_id` e `in_reply_to_user_id`**

In [124]:
df_api_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1746 entries, 0 to 1819
Data columns (total 10 columns):
tweet_id                 1746 non-null int64
created_at               1746 non-null datetime64[ns, UTC]
in_reply_to_status_id    22 non-null float64
in_reply_to_user_id      22 non-null float64
is_quote_status          1746 non-null bool
retweet_count            1746 non-null int64
favorite_count           1746 non-null int64
jpg_url                  1746 non-null object
retweeted                1746 non-null bool
favorited                1746 non-null bool
dtypes: bool(3), datetime64[ns, UTC](1), float64(2), int64(3), object(1)
memory usage: 114.2+ KB


------
<a id="armazenamento"></a>

# Armazenamento

# Relatórios

* Data wrangling efforts
* Analyses and visualizations